### PR TITLE
LLM: GPU QLoRA update to bf16 to accelerate gradient checkpointing

### PR DESCRIPTION
**Pull request by rnwang04**
### 1. Why the change?

https://github.com/analytics-zoo/nano/issues/775#issuecomment-1820116754

### 2. User API changes

No change.

### 3. Summary of the change 

- Update `LoraLowBitLinear` to force input to bf16 on GPU (to solve gradient checkpointing issue)
- Update related GPU QLoRA examples to cast lora Linear to bf16 (to solve gradient checkpointing issue and reduce GPU memory use in all cases)

### 4. How to test?
- [ ] Unit test
- [x] Local test

### CODE REVIEW

File: python/llm/example/GPU/QLoRA-FineTuning/qlora_finetuning.py [hunk 1]
```diff
@@ -61,6 +61,17 @@
...
+
+    for name, module in model.named_modules():
+        if isinstance(module, LoraLayer):
+            module = module.to(torch.bfloat16)
+        if 'norm' in name:
+            module = module.to(torch.float32)
+        if 'lm_head' in name or 'embed_tokens' in name:
+            if hasattr(module, 'weight'):
+                if module.weight.dtype == torch.float32:
+                    module = module.to(torch.bfloat16)
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/qlora_finetuning.py:73]: put these into a utility function

**Review by jason-dai — APPROVED** [review]
LGTM

File: python/llm/example/GPU/QLoRA-FineTuning/qlora_finetuning.py (hunk 1 above)
Comment (rnwang04) [reply python/llm/example/GPU/QLoRA-FineTuning/qlora_finetuning.py]: Have put this logic into `cast_lora_weight` function.
